Fix PR Issues

micgro42 opened the pull request:
The issue with absolute links becoming relative ( https://github.com/michitux/dokuwiki-plugin-move/pull/6/files#diff-bfa8496d4ec28695c5f905a2fe17b1a5R112 ) is addressed in commits 2b7004c...b38aa59

The issues with the locking mentioned in d791aa26159b5c061bbe4fc5328d5b6db9e343d3 of the original PR should be mainly addressed in my commits f677646...e45cf51 . However there are some other fixes still mixed-in.
On a special note: I do not know how to create a new TestRequest that does NOT know about the process-based lock. Hence testing that functionality is somewhat insufficient.

The redundant code ( https://github.com/michitux/dokuwiki-plugin-move/pull/6#discussion-diff-15149339 ) is removed in commit 0bdd57d.

micgro42 committed:
Test that absolute tests remain absolute

micgro42 committed:
Handle relative and absolute links w/ ending colon

micgro42 committed:
When adding .. to relative links, also add a colon

Using the notation with an extra colon as opposed to the shorthand is
clearer
The shorthand is broken for more than one level, see Dokuwiki Issue
 \#1076: https://github.com/splitbrain/dokuwiki/issues/1076

micgro42 committed:
Remove redundant code.

This is already done by the line before: $meta =
$this->getMoveMeta($id);

micgro42 committed:
Do not write if statements on one line.

Writing if statements on one line is bad for code coverage, because the
line will always show up as covered, whether the then statement was
actually executed or note.

micgro42 committed:
Add test to close uncovered case

micgro42 committed:
Create and remove general move-lockfile

micgro42 committed:
Dirty snapshot of a unittest, things are still unclear

Unclear why metadata behaves differently for start and :start

micgro42 committed:
Fix skipping on retry and add possible lock position

micgro42 committed:
Use conf-value for lock-dir and print-outs

micgro42 committed:
Only actual unit tests should add coverage

micgro42 committed:
Add unit test for helper_plugin_move_plan::findAffectedPages

micgro42 committed:
global $PLUGIN_MOVE_WORKING should be unset, not 0

micgro42 committed:
src ns is not affected, it's gone. dst ns is affected

micgro42 committed:
Also check lockfile and remove debug output

Currently there is no use-case for the process-based lock, where the
lockfile alone would not be enough. Maybe the process-based lock can
be removed?

micgro42 committed:
When the page is locked, rewrite it for show only

micgro42 committed:
Test for requests during move

However this does not test for a request request which starts a new
process. So pure process-based locking still lets this test pass.

micgro42 committed:
Remove debug printouts

micgro42 committed:
Make helper_plugin_move_plan::stepThroughDocuments unit-testable

The helper_plugin_move_op class-instance was moved to a class object so
we are able to replace it with an mock object in unit-tests. Doing this
we are both able to break the dependency and prevent any pages from
actually being moved and sense the values handed to the
helper_plugin_move_op instance.

The stepThroughDocuments function has now suffcient unit-test coverage
to begin refactoring.

micgro42 committed:
Refactor stepThroughDocuments to remove GOTOs

micgro42 committed:
Check that the correct values are save to file

micgro42 committed:
Adjust structure to comments and vice versa

micgro42 committed:
Use more elaborate and precise variable names

micgro42 committed:
Ensure that the function has only 1 exit

micgro42 committed:
Save options only once, at the end of the function

micgro42 committed:
Reduce filesytem writes by splitting log function

micgro42 committed:
fix .travis

micgro42 committed:
Test that links are correctly rewritten when moving a namespace

micgro42 committed:
Fix failing rewrites because the page was still locked

The change introduced in c50d2a48e18f43f87725c9a868eba421bb1e6baf caused
the rewrite function to delete the move meta data, even so the change
hasn't yet been written to disk.

micgro42 committed:
Fix findAffectedPages for media files

micgro42 committed:
Mark an incomplete test as incomplete

micgro42 committed:
Check for missing media links

findMissingPages and stepThroughMissingPages were generalized to
findMissingDocuments and stepThroughMissingDocuments and now also expect
the type parameter.

Added a step in nextStep for stepThroughMissingDocuments with media type

$referenceidx is solely used inside findMissingDocuments and hence I
made it a local variable instead of a class variable.

micgro42 committed:
Test move of pages&media, Fix steps and teardown

micgro42 committed:
Group tests which take longer than 1 sec as slow

micgro42 committed:
Add unittests for findMissingDocuments

micgro42 committed:
Test affectedPages with media files

micgro42 committed:
Fix bug in stepThroughNamespaces.This needs more work

micgro42 committed:
Also delete meta directory on teardown

micgro42 committed:
Add tests for moving the subscriptions to namespace and pages

micgro42 committed:
Remove obsolete comments

splitbrain commented on admin/main.php:
```diff
@@ -252,7 +253,7 @@ protected function btn($control, $show = true) {
         if($control == 'retry') {
             $control = 'continue';
-            $skip    = 1;
+            $skip    = 0; //todo: this should be 0 for retry and 1 for skip
```
There's a todo comment here. Is this still open?

splitbrain commented on helper/plan.php:
```diff
@@ -453,57 +463,65 @@ public function previewHTML() {
+            $return_items_run = $this->options[$items_run_counter];
         }
+        $this->write_log($log);
+        $this->saveOptions();
```
you moved saveOptions() outside the loop. When the move operation gets interrupted for whatever reason (timelimit hit, memory, etc...) it will pick up at the wrong place.

splitbrain commented on helper/plan.php:
```diff
@@ -562,10 +591,14 @@ protected function stepThroughMissingPages() {
 
-            $Rewriter->setMoveMetas($page, $miss, 'pages');
+            if ($type == self::TYPE_PAGES) {
+                $Rewriter->setMoveMetas($page, $miss, 'pages');
```
should this use the TYPE_PAGES constant?

splitbrain commented on helper/plan.php:
```diff
@@ -681,22 +715,38 @@ protected function addToAffectedPagesList($pages) {
         $pages = array();
         if($type == self::TYPE_PAGES) {
-            $pages = $idx->lookupKey('relation_references', $src);
+            $pages = $idx->lookupKey('relation_references', $src_);
```
$src_ might be undefined

micgro42 replied:
fixed in ca74e8f

splitbrain commented on helper/plan.php:
```diff
@@ -705,33 +755,39 @@ protected function findAffectedPages($src, $class, $type) {
-            } else {
-                $this->referenceidx = file($fn, FILE_IGNORE_NEW_LINES);
-            }
+    protected function findMissingDocuments($src, $dst, $type = self::TYPE_PAGES) { //FIXME: expects $src without colon, but $dst with colon
```
re the FIXME. any way to either check for correct input or fix it automatically?

splitbrain commented on helper/rewrite.php:
```diff
@@ -133,7 +146,16 @@ public static function isLocked() {
         global $PLUGIN_MOVE_WORKING;
+        global $conf;
         $PLUGIN_MOVE_WORKING = $PLUGIN_MOVE_WORKING ? $PLUGIN_MOVE_WORKING + 1 : 1;
+        $lockfile = $conf['lockdir'] . 'move.lock';
```
the lock file name would conflict with a lock for the page "move". I suggest using "_plugin_move.lock"

micgro42 committed:
Fix occasionally undefined variable

micgro42 committed:
Fix rewriting of links to moved mediafiles starting with :

micgro42 committed:
Save wip

Commit:
Fix test to not depend on linkbase detection

Commit:
Fix wrong affected pages by media move

Commit:
Fix findMissingDocuments. page_exists doesn't work for media

micgro42 committed:
Delete all data to fix random fails

micgro42 committed:
Skip broken test

micgro42 committed:
Remove fixed todo comment

micgro42 committed:
findMissingDocuments expects both src and dst to end with colon

micgro42 committed:
Save to file after every move

This impacts performance but should improve recovery when a move is
interrupted

micgro42 committed:
Better filename for lockfile and saver writing
